src/tests/functional/form_submission_tests.ts

```diff
@@ -82,7 +82,6 @@ export class FormSubmissionTests extends TurboDriveTestCase {
 
     await this.nextEventNamed("turbo:before-visit")
     await this.nextEventNamed("turbo:visit")
-    await this.nextEventNamed("turbo:before-cache")
```

Comment:
Is this assertion omitted because we're no longer firing the `turbo:before-cache` event from submissions? 

That's a breaking change to me, since the documentation encourages applications to [listen for `turbo:before-cache`](https://turbo.hotwired.dev/handbook/building#understanding-caching) events to deconstruct any DOM extensions and plugins.

Could we continue to fire the event, but handle the snapshot caching internally?

I'm imagining a less intrusive diff that re-uses some of the already existent `VisitOptions` keys. 

It could extend the `visitCachedSnapshot(Snapshot)` callback to also accept the `SnapshotCache` instance, then clear the cache after a destructive form submission:

```diff
diff --git a/src/core/drive/navigator.ts b/src/core/drive/navigator.ts
index 59b2ad0..8fdccbb 100644
--- a/src/core/drive/navigator.ts
+++ b/src/core/drive/navigator.ts
@@ -6,6 +6,8 @@ import { expandURL, getAnchor, getRequestURL, Locatable, locationIsVisitable } f
 import { getAttribute } from "../../util"
 import { Visit, VisitDelegate, VisitOptions } from "./visit"
 import { PageSnapshot } from "./page_snapshot"
+import { Snapshot } from "../snapshot"
+import { SnapshotCache } from "./snapshot_cache"
 
 export type NavigatorDelegate = VisitDelegate & {
   allowsVisitingLocationWithAction(location: URL, action?: Action): boolean
@@ -85,13 +87,15 @@ export class Navigator {
     if (formSubmission == this.formSubmission) {
       const responseHTML = await fetchResponse.responseHTML
       if (responseHTML) {
+        let visitCachedSnapshot = (snapshot: Snapshot, snapshotCache: SnapshotCache) => {}
         if (formSubmission.method != FetchMethod.get) {
           this.view.clearSnapshotCache()
+          visitCachedSnapshot = (snapshot, snapshotCache) => snapshotCache.clear()
         }
 
         const { statusCode, redirected } = fetchResponse
         const action = this.getActionForFormSubmission(formSubmission)
-        const visitOptions = { action, response: { statusCode, responseHTML, redirected } }
+        const visitOptions = { action, response: { statusCode, responseHTML, redirected }, visitCachedSnapshot }
         this.proposeVisit(fetchResponse.location, visitOptions)
       }
     }
diff --git a/src/core/drive/visit.ts b/src/core/drive/visit.ts
index 9e900b8..3bf4dc2 100644
--- a/src/core/drive/visit.ts
+++ b/src/core/drive/visit.ts
@@ -4,6 +4,7 @@ import { FetchResponse } from "../../http/fetch_response"
 import { History } from "./history"
 import { getAnchor } from "../url"
 import { Snapshot } from "../snapshot"
+import { SnapshotCache } from "./snapshot_cache"
 import { PageSnapshot } from "./page_snapshot"
 import { Action } from "../types"
 import { uuid } from "../../util"
@@ -43,7 +44,7 @@ export type VisitOptions = {
   referrer?: URL,
   snapshotHTML?: string,
   response?: VisitResponse
-  visitCachedSnapshot(snapshot: Snapshot): void
+  visitCachedSnapshot(snapshot: Snapshot, snapshotCache: SnapshotCache): void
   willRender: boolean
 }
 
@@ -73,7 +74,7 @@ export class Visit implements FetchRequestDelegate {
   readonly action: Action
   readonly referrer?: URL
   readonly timingMetrics: TimingMetrics = {}
-  readonly visitCachedSnapshot: (snapshot: Snapshot) => void
+  readonly visitCachedSnapshot: (snapshot: Snapshot, snapshotCache: SnapshotCache) => void
   readonly willRender: boolean
 
   followedRedirect = false
@@ -396,7 +397,7 @@ export class Visit implements FetchRequestDelegate {
 
   cacheSnapshot() {
     if (!this.snapshotCached) {
-      this.view.cacheSnapshot().then(snapshot => snapshot && this.visitCachedSnapshot(snapshot))
+      this.view.cacheSnapshot().then(snapshot => snapshot && this.visitCachedSnapshot(snapshot, this.view.snapshotCache))
       this.snapshotCached = true
     }
   }
```

What do you think?

Reply:
> Is this assertion omitted because we're no longer firing the turbo:before-cache event from submissions?
> 
> That's a breaking change to me, since the documentation encourages applications to listen for turbo:before-cache events to deconstruct any DOM extensions and plugins.

Correct.  Seems odd to fire a `turbo:before-cache` event when it's not going to cache.  If the page includes `<meta name="turbo-cache-control" content="no-cache">`, then turbo does not send a `turbo:before-cache` event.  Would this then be reported as a bug if `turbo:before-cache` is fired, but not actually cached?

Reply:
Seems correct to me that we wouldn't fire `turbo:before-cache` if we indeed no longer cache the page on form submissions? CC @seanpdoyle